config/store.go

```diff
@@ -248,9 +248,28 @@ func (s *Store) loadLockedWithOld(oldCfg *model.Config, unlockOnce *sync.Once) e
 	if err != nil {
 		return errors.Wrap(err, "failed to marshal loaded config")
 	}
-	if !s.readOnly && (len(configBytes) == 0 || !bytes.Equal(oldCfgBytes, newCfgBytes)) {
-		if err := s.backingStore.Set(s.configNoEnv); err != nil {
-			if !errors.Is(err, ErrReadOnlyConfiguration) {
+
+	var shouldStore bool
+	hasChanged := len(configBytes) == 0 || !bytes.Equal(oldCfgBytes, newCfgBytes)
+	if hasChanged {
+		featureFlags := s.configNoEnv.FeatureFlags
+		// Don't persist feature flags unless we are on MM cloud
+		// MM cloud uses config in the DB as a cache of the feature flag
+		// settings in case the management system is down when a pod starts.
+		if !s.persistFeatureFlags {
+			s.configNoEnv.FeatureFlags = nil
+		}
+		toStoreBytes, err := json.Marshal(s.configNoEnv)
+		if err != nil {
+			return errors.Wrap(err, "failed to marshal old config")
+		}
+		shouldStore = !bytes.Equal(toStoreBytes, configBytes)
```

Comment:
Wondering if it's possible to check if the `FeatureFlags` wasn't nil previously and was set to nil, and use that as boolean result for `shouldStore` rather than doing a `json.Marshal` and a `bytes.Equal`?

Reply:
That thought crossed my mind. The `shouldStore` isn't only about `FeatureFlags` though. Again the approach here was very conservative and I opted to repeat some logic rather than possibly breaking something (to be properly reviewed in [MM-32390](https://mattermost.atlassian.net/browse/MM-32390) of course).

Mostly I want to make sure that what we read from the backing store (`configBytes`) is different from what we are going to write (`s.configNoEnv`) to avoid triggering the set. The db store actually does this already internally but I am thinking a `json.Marshal` + `bytes.Equal` while not ideal should be more efficient (at least time-wise) than a db query or reading a file.

That said, I have a feeling a lot of this logic can be simplified (and possibly removed) but since it's just a feeling I am opting for the dead simple approach of actually checking the data (even if twice).

Reply:
Gotcha. Overall, the `Set` flow looks very complicated and hard to read for me. I thought the existing json.Marshal comparison was for that purpose. But it looks like it's not? That seems to compare `newCfgBytes, err := json.Marshal(loadedConfig)`, but it actually writes `s.configNoEnv`. IMO, there should only be 2 `json.Marshal`s, not 3. 

There seems to be 3 sets of configs that we are comparing:
- The old config struct
- The old config loaded from the backing store
- The new config that's going to be written.

And then there are multiple filters and environment maps being applied to each of them.

Totally agree that a lot of this logic is redundant and needs to be cleaned up. So will be looking forward to MM-32390 :)

Reply:
Yes, I didn't really trust all those transformations hence the added serialization + check on `configNoEnv` but I am confident that when I get to refactor this we should be able to avoid several checks and limit the marshaling to a minimum.